src/Lib/BSP/Lumps/BspLumps/EntityLump.cs

```diff
@@ -112,28 +113,42 @@ private bool ReadEntity(BinaryReader reader)
                     "WARNING: Failed to parse entity: {0}, {1} in list.\n Saving this BSP could cause data loss!",
                     e.Message, Data.Count);
 
-                // Read to end of entity (barf)
-                for (var c = ' '; c != '\0' && c != '}'; c = reader.ReadChar())
+                // Read to end of entity
+                bool foundEnd = false;
+                while (reader.BaseStream.Position < endPos)
                 {
+                    var c = reader.ReadChar();
+                    if (c == '\0' || c == '}')
+                    {
+                        foundEnd = true;
+                        break;
+                    }
                 }
+                if (!foundEnd)
+                    Console.WriteLine("WARNING: End of entity not found!");
             }
 
             return false;
         }
 
-        public override void Write(BinaryWriter writer)
+        public override void Write(Stream stream)
         {
             foreach (var ent in Data)
             {
-                writer.Write("{");
-                foreach (var (key, value) in ent.Properties)
-                    writer.Write($"\"{key}\" \"{value}\"");
-                foreach (var (key, value) in ent.IOProperties) 
-                    writer.Write($"\"{key}\" \"{value.TargetEntityName},{value.Input},{value.Parameter},{value.Delay},{value.TimesToFire}\"");
-                writer.Write("}\0");
+                stream.Write(Encoding.ASCII.GetBytes("{\n"));
```

Comment:
How long does `Encoding.ASCII.GetBytes` take to execute? Should the static ones (all but the `${prop}` one) be cached outside of this foreach loop?

Reply:
I would hope the compiler/optimization understand that the output of this is always the same and not call this function every time. In fact it shouldn't be called at all during runtime since the ASCII value for a fixed string shouldn't change (but maybe you could override something and it can't do this at compile time for some reason).
Even if it calls it every time it shouldn't take long. I can check if you want to but I wouldn't worry about this for now. There are a lot more performance improvements to be made like asynchronous file reading and processing lumps in parallel. Not saying small things like this aren't important :)

Reply:
I completely agree, I'm fine with this for now!